RFC: Reviewdog Diagnostic Format

haya14busa opened the pull request:
This is a proposal of ~Reviewdog Diagnostic Protocol (RDP)~ Reviewdog Diagnostic Format as standard
diagnostic format.
Any review, suggestion, feedback, criticism, and comments from anyone are very
much welcome.

Preview: https://github.com/reviewdog/reviewdog/tree/rdp/proto/rdf

- [x] Updated Unreleased section in [CHANGELOG](https://github.com/reviewdog/reviewdog/blob/master/CHANGELOG.md) or it's not notable changes.

kayoub5 commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,286 @@
+https://github.com/reviewdog/reviewdog repository, but we may create a separate
+repository once it's reviewed and stabilized.
+
+# Reviewdog Diagnostic Protocol (RDP)
```
The abbreviation `RDP` will create ambiguity with the popular `Remote Desktop Protocol`.
I think it would be better to find another abbreviation.

haya14busa replied:
True. Do you have any suggestions? 
I wonder whether *protocol* is good or not as well. https://github.com/reviewdog/reviewdog/tree/rdp/proto/rdp#open-questions

But if we use *Format* instead of *Protocol*, **RDF** (*Reviewdog Diagnostic Format*) is also used by [Resource Description Framework](https://www.w3.org/RDF/) as well.

The acronym is basically for providing short format name (currently rdpjsonl / rdpjson), so maybe it's ok even if there are other usages. Any thought?

jsoref replied:
I'd rather not collide w/ RDF / RDP -- I use both.

The lazy answer would be RDDP or RDDF.

kayoub5 commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,286 @@
+        "range": {
+          "start": {
+            "line": 14,
+            "column": 15
```
Some compilers/linters can only provide absolute position in the file, not line/column.

haya14busa replied:
Yeah, that's actually a good discussion point. The reason why I didn't include the byte offset (absolute position in the file) is that if we added it, all consumers tools of RDP output (such as reviewdog) need to add support to convert byte offset <-> line/column. 

My current idea is the conversion should happen outside RDP, such as in linters or converters.

kayoub5 commented:
@haya14busa Have you considered using an existing protocol? such as LSP?
https://microsoft.github.io/language-server-protocol/specifications/specification-current/#diagnostic

haya14busa commented:
> Have you considered using an existing protocol? such as LSP?

Considered, but no.
See https://github.com/reviewdog/reviewdog/tree/rdp/proto/rdp#language-server-protocol-lsp

kayoub5 commented:
@haya14busa See also
* https://docs.sonarqube.org/latest/analysis/generic-issue/
* https://coala.io/ and http://api.coala.io/en/latest/Developers/Writing_Linter_Bears.html

kayoub5 commented on proto/rdp/reviewdog.proto:
```diff
@@ -0,0 +1,192 @@
+// - rdpjsonl: JSON Lines (http://jsonlines.org/) of the `Diagnostic` message.
+// - rdpjson: JSON format of the `DiagnosticResult` message.
+
+syntax = "proto3";
```
I think using a JSON Schema file to describe the format would be easier for others to use and validate their output

haya14busa replied:
I prefer Protocol Buffer (as primary IDL) because...

- Protobur can auto-generate code for a lot of languages including Go which is used in reviewdog.
- It's more readable (I believe) and I think it's actually not difficult to understand. It supports comments for example.

But, yes that's a good point. 
It should be able to auto-generate json schema from protobuf so we can give it a try. https://github.com/chrusty/protoc-gen-jsonschema

haya14busa replied:
Added. 3e7cb51d7ea1a608c0869052054231e3c72ffade

haya14busa commented:
I happened to find Static Analysis Results Interchange Format (SARIF) https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html

haya14busa commented:
I found SARIF 6 hours ago by finding [@microsoft/eslint-formatter-sarif](https://www.npmjs.com/package/@microsoft/eslint-formatter-sarif) while investigating eslint custom formatter and I just found that GitHub's new Code Scanning (beta) feature uses SARIF for third-party code scanning support. https://help.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/about-code-scanning#about-third-party-code-scanning-tools (I got an email just 40 min ago!)
![Screenshot 2020-06-16 at 03 50 00](https://user-images.githubusercontent.com/3797062/84694639-7b748880-af84-11ea-88c7-cb8f09746ee7.png)

This is very interesting. My proposal can be still useful for some use cases like reviewdog or as a light weight format but discarding my proposal and adopting SARIF can be plausible alternative option.
I'll investigate it more and update the document.

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+
+Any review, suggestion, feedback, criticism, and comments from anyone is very
+much welcome. Please leave comments in Pull Request ([#629](https://github.com/reviewdog/reviewdog/pull/629))
+, in issue [#628](https://github.com/reviewdog/reviewdog/issues/628) or
```
Splitting the comma from the link renders poorly.

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+, in issue [#628](https://github.com/reviewdog/reviewdog/issues/628) or
+file [an issue](https://github.com/reviewdog/reviewdog/issues).
+
+The document and the actual definition are currently in under
```
```suggestion
The document and the actual definition are currently under the
```

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+# Reviewdog Diagnostic Protocol (RDP)
+
+Reviewdog Diagnostic Protocol (RDP. Or **R**eview**D**og **P**rotocol in short.)
+defines standard machine-readable message structures which represents a
```
```suggestion
defines standard machine-readable message structures which represent a
```

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+## Wire formats of Reviewdog Diagnostic Protocol.
+
+RDP uses [Protocol Buffer](https://developers.google.com/protocol-buffers) to
+define the message structure, but recommended wire format is JSON considering
```
```suggestion
define the message structure, but the recommended wire format is JSON considering
```

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+          "text": "<replacement text>"
+        }
+      ],
+      "sevrity": "WARNING"
```
Is this misspelling intentional?

(At the very least my spell checker would scream.)

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+
+Problems:
+- No support for code suggestions (also known as auto-correct or fix).
+- Output format of matched results are undocumented and it seems to be used internally in VSCode nor GitHub Actions.
```
`nor` here doesn't parse -- not sure what you're trying to write :-(

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+Problems:
+- No support for code suggestions (also known as auto-correct or fix).
+- Output format of matched results are undocumented and it seems to be used internally in VSCode nor GitHub Actions.
+- It's hard to write problem matcher for complicated output.
```
```suggestion
- It's hard to write problem matchers for complicated output.
```

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+
+Problems:
+- LSP and the Diagnostic message is basically per one file. It's not always
+  suited to be used as diagnostic tools output because they often needs to
```
```suggestion
  suited to be used as diagnostic tools output because they often need to
```

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+
+### Diagnostic tools' RDP Support
+Ideally, diagnostic tools themselves should support outputing their results as
+RDP compriant format, but not all tools does support RDP especially in early
```
```suggestion
RDP compliant format, but not all tools does support RDP especially in early
```
(I need to file a thing to add my checker...)

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+### Consumer: RDP Report Formatter
+*Not implemented yet*
+
+There are many diagnostic output format (report format) and each diagnostic
```
```suggestion
There are many diagnostic output formats (report formats) and each diagnostic
```

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+*Not implemented yet*
+
+There are many diagnostic output format (report format) and each diagnostic
+tool implement them on their own. e.g. [eslint](https://eslint.org/docs/user-guide/formatters)
```
```suggestion
tool implements them on their own. e.g. [eslint](https://eslint.org/docs/user-guide/formatters)
```

jsoref commented on proto/rdp/README.md:
```diff
@@ -0,0 +1,277 @@
+
+## Reviewdog Diagnostic Protocol Concept
+Again, the idea behind the Reviewdog Diagnostic Protocol (RDP) is to standardize
+the protocol for how diagnostic tools (e.g. compilers, linters, etc..) and
```
In addition to the language tool complaint, as a general rule you should not do `e.g. ..., etc.` -- `e.g.` is used to introduce examples, but `etc.` (*[et cetera](https://dictionary.cambridge.org/dictionary/english/etc)*) is for *and others*, which aren't really *an example*.

Here, the sample solution is to drop ` etc..` entirely. Possibly adding `and` or `or` before `code review API`.
Note: there's no need for the trailing `.` in this, as the parenthetical isn't a sentence.

haya14busa commented:
Thank you everyone for the reviews so far!

I updated the document. Renamed to Reviewdog Diagnostic Format (RD format) and added a section about the SARIF.

I'm going to merge this PR because new upcoming features (like supporting suggestion) depends on it.
However, I'm still welcome any reviews and feedback.
Please feel free to leave any comments here or #628.
